File: frontend/src/api/statistics.ts
import { api } from './client';

function withQuery(path: string, params?: Record<string, string | number | boolean | undefined | null>) {
  const q = new URLSearchParams();
  if (params) {
    for (const [key, value] of Object.entries(params)) {
      if (value === undefined || value === null || value === '') continue;
      q.set(key, String(value));
    }
  }
  const suffix = q.toString() ? `?${q}` : '';
  return `${path}${suffix}`;
}

export type StatsParams = Record<string, string | number | boolean | undefined | null>;

export type ManagerActionBody = {
  action_type: string;
  responsible_manager?: string;
  due_at?: string;
  comment?: string;
  priority?: string;
};

export type ExportReportBody = {
  report_type: string;
  period_from?: string;
  period_to?: string;
  job_id?: string;
  fmt?: string;
  options?: Record<string, unknown>;
};

export const statisticsApi = {
  managerDashboard: (params?: StatsParams) =>
    api.get<Record<string, unknown>>(withQuery('/api/sender/manager-dashboard', params)),

  campaigns: (params?: StatsParams) =>
    api.get<Record<string, unknown>>(withQuery('/api/sender/campaigns', params)),

  recipients: (params?: StatsParams) =>
    api.get<Record<string, unknown>>(withQuery('/api/sender/recipients', params)),

  recipientDetail: (rowKey: string) =>
    api.get<Record<string, unknown>>(`/api/sender/recipients/${encodeURIComponent(rowKey)}`),

  saveRecipientAction: (rowKey: string, body: ManagerActionBody) =>
    api.post<Record<string, unknown>>(
      `/api/sender/recipients/${encodeURIComponent(rowKey)}/action`,
      body,
    ),

  consents: (params?: StatsParams) =>
    api.get<Record<string, unknown>>(withQuery('/api/sender/consents', params)),

  problems: (params?: StatsParams) =>
    api.get<Record<string, unknown>>(withQuery('/api/sender/email-problems', params)),

  campaignAnalytics: (jobId: string, params?: StatsParams) =>
    api.get<Record<string, unknown>>(
      withQuery(`/api/sender/campaign-analytics/${encodeURIComponent(jobId)}`, params),
    ),

  reports: (params?: StatsParams) =>
    api.get<Record<string, unknown>>(withQuery('/api/sender/reports', params)),

  exportReport: (body: ExportReportBody) =>
    api.post<{ report_id?: string; download_url?: string } & Record<string, unknown>>(
      '/api/sender/reports/export',
      body,
    ),

  reportDownloadUrl: (reportId: string) =>
    `/api/sender/reports/download/${encodeURIComponent(reportId)}`,

  deliveryReportUrl: (jobId: string) =>
    `/api/download/sender-delivery-report?job_id=${encodeURIComponent(jobId)}`,

  autoCallContactsUrl: (jobId: string) =>
    `/api/download/auto-call-contacts?job_id=${encodeURIComponent(jobId)}`,
};

File: src/mcp_server/tools/statistics.py
from __future__ import annotations

from typing import Any, Callable


def _stats_params(
    *,
    period_from: str | None = None,
    period_to: str | None = None,
    job_id: str | None = None,
    q: str | None = None,
    limit: int | None = None,
    offset: int | None = None,
    extra: dict[str, Any] | None = None,
) -> dict[str, Any]:
    params: dict[str, Any] = {
        "period_from": period_from,
        "period_to": period_to,
        "job_id": job_id,
        "q": q,
        "limit": limit,
        "offset": offset,
    }
    if extra:
        params.update(extra)
    return params


def register_statistics_tools(mcp: Any, get_client: Callable[[], Any]) -> None:
    @mcp.tool()
    def get_manager_dashboard(
        period_from: str | None = None,
        period_to: str | None = None,
        job_id: str | None = None,
    ) -> Any:
        """Manager dashboard KPIs and charts (`/api/sender/manager-dashboard`)."""
        return get_client().get(
            "/api/sender/manager-dashboard",
            params=_stats_params(period_from=period_from, period_to=period_to, job_id=job_id),
        )

    @mcp.tool()
    def list_stats_campaigns(
        period_from: str | None = None,
        period_to: str | None = None,
        q: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> Any:
        """List campaigns in statistics view (`/api/sender/campaigns`)."""
        return get_client().get(
            "/api/sender/campaigns",
            params=_stats_params(
                period_from=period_from,
                period_to=period_to,
                q=q,
                limit=limit,
                offset=offset,
            ),
        )

    @mcp.tool()
    def list_stats_recipients(
        period_from: str | None = None,
        period_to: str | None = None,
        job_id: str | None = None,
        q: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> Any:
        """List recipients in statistics view."""
        return get_client().get(
            "/api/sender/recipients",
            params=_stats_params(
                period_from=period_from,
                period_to=period_to,
                job_id=job_id,
                q=q,
                limit=limit,
                offset=offset,
            ),
        )

    @mcp.tool()
    def get_stats_recipient(row_key: str) -> Any:
        """Get recipient detail by row_key."""
        return get_client().get(f"/api/sender/recipients/{row_key}")

    @mcp.tool()
    def save_stats_recipient_action(row_key: str, body: dict[str, Any]) -> Any:
        """Save a manager action on a recipient. Body: action_type, responsible_manager, due_at, comment, priority."""
        return get_client().post(f"/api/sender/recipients/{row_key}/action", body)

    @mcp.tool()
    def list_stats_consents(
        period_from: str | None = None,
        period_to: str | None = None,
        q: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> Any:
        """List consent records in statistics view."""
        return get_client().get(
            "/api/sender/consents",
            params=_stats_params(
                period_from=period_from,
                period_to=period_to,
                q=q,
                limit=limit,
                offset=offset,
            ),
        )

    @mcp.tool()
    def list_email_problems(
        period_from: str | None = None,
        period_to: str | None = None,
        q: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> Any:
        """List email delivery problems."""
        return get_client().get(
            "/api/sender/email-problems",
            params=_stats_params(
                period_from=period_from,
                period_to=period_to,
                q=q,
                limit=limit,
                offset=offset,
            ),
        )

    @mcp.tool()
    def get_campaign_analytics(
        job_id: str,
        period_from: str | None = None,
        period_to: str | None = None,
    ) -> Any:
        """Campaign analytics for a job_id."""
        return get_client().get(
            f"/api/sender/campaign-analytics/{job_id}",
            params=_stats_params(period_from=period_from, period_to=period_to),
        )

    @mcp.tool()
    def list_reports(
        period_from: str | None = None,
        period_to: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> Any:
        """List generated statistics reports."""
        return get_client().get(
            "/api/sender/reports",
            params=_stats_params(
                period_from=period_from,
                period_to=period_to,
                limit=limit,
                offset=offset,
            ),
        )

    @mcp.tool()
    def export_report(body: dict[str, Any]) -> Any:
        """Export a statistics report. Body: report_type, period_from, period_to, job_id, fmt, options."""
        return get_client().post("/api/sender/reports/export", body)

    @mcp.tool()
    def get_report_download_meta(report_id: str) -> Any:
        """Fetch report download metadata/bytes as base64 (`/api/sender/reports/download/{id}`)."""
        return get_client().request(
            "GET",
            f"/api/sender/reports/download/{report_id}",
            raw=True,
        )

    @mcp.tool()
    def get_domain_delivery_stats(
        period_from: str | None = None,
        period_to: str | None = None,
        job_id: str | None = None,
    ) -> Any:
        """Domain delivery statistics."""
        return get_client().get(
            "/api/sender/domain-delivery-stats",
            params=_stats_params(period_from=period_from, period_to=period_to, job_id=job_id),
        )
